### ISSUE TITLE
Opening a patch with file association deletes autosave patch with no warning

### ISSUE THREAD

**Issue by jonheal** [VCVRack/Rack#762]
This one has bit me many times.

The autosave is great, but misleading. When I open Rack, the previous project I was working on appears to open as its name appears in the title bar. I make changes and quit. Later, I reopen, and sure enough, the most recent changes are there, intact.

However, if I create a new rack with the New menu command, work for a while, and then at some point decide to open my previous project, when I do so, I find none of my recent changes to that project are there (unless I explicitly invoked the Save menu command). The autosave function has, at worse, mislead me to believe I was automatically writing changes to my file when I was actually writing changes to the autosave file, or at best, made me complacent.

I suggest one of two things, ether:
1. Merge the autosave.vcv file with the file I am working on at application end, or;
2. Change the new file dialog to a "save as ..." dialog.

**Comment by AndrewBelt** [comment]
Would a more descriptive "Are you sure you want to start a new patch?" dialog fix it? And/or a "Are you sure you want to overwrite the current patch?" when opening a patch?

I think that it's fairly intuitive that an explicit save is required before starting a new patch or opening it. Just like Microsoft Word (or really any other document-based software), you need to save before clearing or opening another document, while the current memory holds the current state. The *only* difference between that and Rack is closing and reopening behavior.

I'll wait for others' :+1:'s on this issue and come back to it later.

**Comment by jonheal** [comment]
The difference with Word, in particular, is that creating a new document does not close/overwrite your current unsaved document, it invokes another instance of Word.

Yes, changing the wording of the dialog could help. I would also add "You will lose unsaved changes." Also, maybe do like Reason does, and put an asterisk after the file name in the title bar if it includes unsaved changes.

**Comment by AndrewBelt** [comment]
Okay, maybe Ableton Live or REAPER is a better example.

Right now it's impossible to know whether a patch has been changed, so can't use an "*" until #153

**Comment by jonheal** [comment]
Or can invoking the New file command while an existing, previously saved project is open, also make a call the Save function in the background to save changes to the existing file?

**Comment by AndrewBelt** [comment]
>make a call the Save function in the background to save changes to the existing file

No, most of the time you **don't** want to do that.

**Comment by jonheal** [comment]
Well, I would. But you're the boss. I will adapt to your way of thinking on this and try to get in the habit of explicitly saving more often.

--
Jon Heal
http://www.theheals.org/
Have a nice day!

**Comment by DomiKamu** [comment]
Module deletion from rack while autosaving (either installable 0.5.1 or 0.6.0dev) may cause "memory segmentation" issues (and crashes). Experimented with Core / Fundamental / Befaco modules, also third party modules such Blamsoft XFX, VCV Console/PM and Vult.

**Comment by ZwergNaseXXL** [comment]
Was as well bitten by autosave as @jonheal describes. My suggestion:
- Keep autosaving as is, no matter if working on a new/unnamed/unsaved rack or on a named/saved one.
- Whenever starting a new rack or closing the app or loading a saved rack and the current patch wasn't saved, prompt users if they want to save their work. When they answer no, just continue with the requested operation. When they answer yes, in case of unnamed work display the "save as..." dialog, in case of named work just save to that file.
I think a "do you want to save" box is more useful than a "are you sure you want to start over" box, which also covers only one of the cases where work might be lost.

**Comment by AndrewBelt** [comment]
I'm not a fan of launching/closing dialog messages. My hope is to get rid of all of them and put them in "info notification" boxes after launch so the user has zero friction when starting/stopping a patch. The nice thing about Eurorack is the ability to turn it on and start playing, and then turn it off and come back to it later.

My advice is to just get used to it. Hitting Ctrl+S is no new habit in software. There's no better way that satisfies everyone.

**Comment by AndrewBelt** [comment]
When undo history #153 is implemented, I'll be able to add a "*" icon to the title, which will be preserved across Rack launches.

**Comment by jonheal** [comment]
Already used to it. Close if you want.

--
Jon Heal
IT Specialist, Information Services Division - Business Solutions Management Branch
United States Department of Agriculture | Foreign Agricultural Service | Office of the Chief Operating Officer
202-720-5728
Have a nice day

**Comment by jonheal** [comment]
I didn’t know you were planning an undo history. Convenient, for sure. For myself, I am trying to be less anal and possessive and embrace the whole modular “here today, gone tomorrow” aesthetic.

--
Jon Heal
IT Specialist, Information Services Division - Business Solutions Management Branch
United States Department of Agriculture | Foreign Agricultural Service | Office of the Chief Operating Officer
202-720-5728
Have a nice day

**Comment by AndrewBelt** [comment]
Adding knowledge of a modified state will allow me to intelligently handle unsaved work. Once this is in place, I can do the following things:
- Add a "*" character to the filename like Microsoft Word, etc.
- Add "Would you like to save your patch XXX.vcv" when clicking New or Open.

I think that would completely solve this, correct?

**Comment by ZwergNaseXXL** [comment]
Yes, that's exactly what I asked for. Also on closing the app, please.

**Comment by AndrewBelt** [comment]
Closing since this is solved.

**Comment by jonheal** [comment]
This is an aside and not really your issue, but a “gotcha” occurs if you have been working on a patch, but not actually saved it to a separate file. Then you go to the forum and download and open some one’s damn patch they’re shared. Bye bye, patch you were working on but neglected to save.

**Comment by AndrewBelt** [comment]
Okay, I have reopened and renamed the issue to something more descriptive.
